web/auth_test.go

```diff
@@ -0,0 +1,204 @@
+
+// +build root, integration
+
+package web
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"os/user"
+	"testing"
+	"github.com/zenoss/glog"
+)
+
+var (
+	ErrTestUserExists = errors.New("test user exists. could not create new")
+)
+
+const (
+	testUserPassword = "secret"
+	testUserPasswordEncrypted = "$1$Qd8H95T5$RYSZQeoFbEB.gS19zS99A0"
+	testUserSalt = "$1$Qd8H95T5$"
+)
+
+type testUser struct {
+	username string
+	password string
+	group    string
+	expired  bool
+}
+
+var testUsers = map[string]testUser{
+	"gooduser": testUser{"ztestgooduser", "ztestgooduserpass", adminGroup, false},
+	"nonrootuser": testUser{"ztestplainuser", "ztestplainuserpass", "users", false},
+	"oldbutgooduser": testUser{"ztestolduser", "ztestolduserpass", adminGroup, true},
+}
+
+var createdUsers = make(map[string]testUser)
+
+type testCase struct {
+	user                testUser
+	testPassword        string
+	testGroup           string
+	expectedPamResult   bool
+	expectedGroupResult bool
+	expectedResult      bool
+	description         string
+}
+
+var testCases = []testCase{
+	testCase{
+		user:                testUsers["gooduser"],
+		testPassword:        testUsers["gooduser"].password,
+		testGroup:           adminGroup,
+		expectedPamResult:   true,
+		expectedGroupResult: true,
+		expectedResult:      true,
+		description:         "good user with correct password should pass PAM, pass group check, and should validate",
+	},
+	testCase{
+		user:                testUsers["gooduser"],
+		testPassword:        "",
+		testGroup:           adminGroup,
+		expectedPamResult:   false,
+		expectedGroupResult: true,
+		expectedResult:      false,
+		description:         "good user with empty password should fail PAM, pass group check, and should not validate.",
+	},
+	testCase{
+		user:                testUsers["nonrootuser"],
+		testPassword:        testUsers["nonrootuser"].password,
+		testGroup:           adminGroup,
+		expectedPamResult:   true,
+		expectedGroupResult: false,
+		expectedResult:      false,
+		description:         "nonroot user should pass PAM, fail authentication, and should not validate.",
+	},
+	testCase{
+		user:                testUsers["oldbutgooduser"],
+		testPassword:        testUsers["oldbutgooduser"].password,
+		testGroup:           adminGroup,
+		expectedPamResult:   false,
+		expectedGroupResult: true,
+		expectedResult:      false,
+		description:         "user with expired login should pass PAM, fail group check, and should not validate.",
+	},
+}
+
+
+func TestMain(m *testing.M) {
+	if 0 != os.Geteuid() {
+		glog.Infof("Must be root to run integration tests. Exiting (no tests run).")
+		os.Exit(0)
+	}
+	err := CreateTestUsers()
+	if err != nil {
+		glog.Errorf("Error creating test user: %s\n", err)
+	}
+	glog.Infof("Running tests - some errors expected from negative tests.\n")
+	result := m.Run()
+	glog.Infof("Removing test users\n")
+	RemoveTestUsers()
+	glog.Infof("Exiting\n")
+	os.Exit(result)
+}
+
+func CreateTestUsers() error {
```

Comment:
nitpick - tecnically CreateTestUsers(), RemoveTestUsers() and RemoveTestUser() should be package-private because they're not needed outside of this file